

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>gwide &mdash; gwide 0.1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="_static/css/main.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="gwide 0.1.0 documentation" href="#"/>
        <link rel="next" title="gwide Reference manual" href="ref.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="#" class="icon icon-home"> gwide
          

          
          </a>

          
            
            
              <div class="version">
                0.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">gwide</a></li>
</ul>
<p class="caption"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="ref.html">gwide Reference manual</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">gwide</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#">Docs</a> &raquo;</li>
        
      <li>gwide</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="gwide">
<h1>gwide<a class="headerlink" href="#gwide" title="Permalink to this headline">¶</a></h1>
<p>(the documentation is under construction, come back later for more)</p>
<p>Python package for downstream analysis of CRAC data. Package developed for analysis of RNA polymerase III data published in
Global analysis of transcriptionally engaged yeast RNA polymerase III reveals extended tRNA transcripts.
Turowski TW, Leśniewska E, Delan-Forino C, Sayou C, Boguta M, Tollervey D.
Genome Res. 2016 May 20. pii: gr.205492.116.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/27206856">http://www.ncbi.nlm.nih.gov/pubmed/27206856</a></p>
</div>
<div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>for users:
.. code:: shell</p>
<blockquote>
<div>pip install &#8211;user mercurial git+git://github.com/tturowski/gwide.git (or skip &#8211;user mercurial to install globally);</div></blockquote>
<p>add to your .profile or .bash_profile or .bash_login:
.. code:: shell</p>
<blockquote>
<div>PATH=&#8221;$HOME/.local/bin:$PATH&#8221;</div></blockquote>
<p>for developers:
.. code:: shell</p>
<blockquote>
<div>git clone <a class="reference external" href="mailto:git&#37;&#52;&#48;github&#46;com">git<span>&#64;</span>github<span>&#46;</span>com</a>:tturowski/gwide.git
pip install &#8211;editable .</div></blockquote>
</div>
<div class="section" id="setup">
<h1>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h1>
<p>It&#8217;s possible to create $HOME/bin/default.aml file with paths to follwing files:
#default paths to genome files
GTF_PATH : <a href="#id1"><span class="problematic" id="id2">*</span></a>.gtf
FASTA_PATH : <a href="#id3"><span class="problematic" id="id4">*</span></a>.fasta
TAB_PATH : <a href="#id5"><span class="problematic" id="id6">*</span></a>.tab</p>
</div>
<div class="section" id="description">
<h1>Description<a class="headerlink" href="#description" title="Permalink to this headline">¶</a></h1>
<p>ploting of genome-wide plots pyCRAC downstream tool</p>
<dl class="docutils">
<dt>gwideHittable (-h for help, input <a href="#id7"><span class="problematic" id="id8">*</span></a>.hittable) - Three options:</dt>
<dd><ul class="first last simple">
<li>Calculate &#8220;correlations&#8221;: Pearson, Spearman or Kendall Tau</li>
<li>&#8220;count&#8221; hittables for further analysis. Ideal to work with multiple experiments</li>
<li>Plot &#8220;piechart&#8221;s for hittable classes (plots are not ideal...)</li>
</ul>
</dd>
<dt>novo2concat.py (-h for help):</dt>
<dd><ul class="first last simple">
<li>copy <a href="#id9"><span class="problematic" id="id10">*</span></a>.novo files to new folder</li>
<li>prefix from <a href="#id11"><span class="problematic" id="id12">*</span></a>.novo file is used as experiment name</li>
</ul>
</dd>
<dt>gwidePlot (-h for help, input <a href="#id13"><span class="problematic" id="id14">*</span></a>.concat):</dt>
<dd><ul class="first last simple">
<li>plot genome wide plots, 5&#8217; and 3&#8217; end aligned or aligned to choosen aligner (-o aligner) or aligneg to 3&#8217; end of read-through</li>
<li>plot genome wide ratio between different experiment</li>
<li>possibility to filter genes using -f option</li>
<li>printing <a href="#id15"><span class="problematic" id="id16">*</span></a>.csv tables to generate heatmaps using other software (i.e. GENE-E)</li>
<li>calculate p-value for a non-canonical termination sites</li>
<li>making GTF files (for whole transcripts or only for extensions</li>
</ul>
</dd>
<dt>gwidetRNA (-h for help, input <a href="#id17"><span class="problematic" id="id18">*</span></a>.concat):</dt>
<dd><ul class="first last simple">
<li>plot single tRNA plots: multiple experiments per page, one experiment per page, one experiment under another, mark A and B boxes</li>
<li>plot ratio between experiments</li>
<li>plot under nucleotides resolution</li>
<li>calculate dG RNA_DNA/DNA_DNA for each valley or last 20 nt after tRNA gene</li>
<li>save tab-deliminated file with each plot calculations</li>
</ul>
</dd>
<dt>GTF parser tools (-h for help, work with standard input):</dt>
<dd><ul class="first last simple">
<li>getFastaSeqs</li>
<li>getGeneLength</li>
<li>getIdFromName</li>
<li>getNameFromId</li>
<li>getGeneNamesFromGTF</li>
<li>getNameFromId4Tab</li>
</ul>
</dd>
</dl>
</div>
<div class="section" id="citation">
<h1>Citation<a class="headerlink" href="#citation" title="Permalink to this headline">¶</a></h1>
<p>Turowski TW, Leśniewska E, Delan-Forino C, Sayou C, Boguta M, Tollervey D.
Global analysis of transcriptionally engaged yeast RNA polymerase III reveals extended tRNA transcripts.
Genome Res. 2016 May 20. pii: gr.205492.116.</p>
<div class="section" id="contribute">
<h2>Contribute<a class="headerlink" href="#contribute" title="Permalink to this headline">¶</a></h2>
<p>gwide is an open-source library originally written at the
Edinburgh Genome Foundry by <a class="reference external" href="https://www.researchgate.net/profile/Tomasz_Turowski">TomaszWTurowski</a>. It is released on <a class="reference external" href="https://github.com/tturowski/gwide">Github</a> under the Apache 2.0
licence (¢ Tomasz W. Turowski), everyone is welcome to contribute.</p>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="ref.html" class="btn btn-neutral float-right" title="gwide Reference manual" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Edinburgh Genome Foundry.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>