

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>gwide Reference manual &mdash; gwide 0.1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="_static/css/main.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="gwide 0.1.0 documentation" href="index.html"/>
        <link rel="prev" title="gwide" href="index.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> gwide
          

          
          </a>

          
            
            
              <div class="version">
                0.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">gwide</a></li>
</ul>
<p class="caption"><span class="caption-text">Reference</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">gwide Reference manual</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#profileanalyser">ProfileAnalyser</a></li>
<li class="toctree-l2"><a class="reference internal" href="#methods">Methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bigwigtools">BigWigTools</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">gwide</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>gwide Reference manual</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/ref.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="gwide-reference-manual">
<h1>gwide Reference manual<a class="headerlink" href="#gwide-reference-manual" title="Permalink to this headline">¶</a></h1>
<div class="section" id="profileanalyser">
<h2>ProfileAnalyser<a class="headerlink" href="#profileanalyser" title="Permalink to this headline">¶</a></h2>
<p>Set of functions designed to analysis of genomic profiles</p>
<span class="target" id="module-gwide.profileAnalyser"></span><dl class="function">
<dt id="gwide.profileAnalyser.compare1toRef">
<code class="descclassname">gwide.profileAnalyser.</code><code class="descname">compare1toRef</code><span class="sig-paren">(</span><em>dataset=Series([]</em>, <em>dtype: float64)</em>, <em>ranges='mm'</em>, <em>heatmap=False</em>, <em>relative=False</em>, <em>reference='/home/tturowski/notebooks/RDN37_reference_collapsed.csv'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/profileAnalyser.html#compare1toRef"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.profileAnalyser.compare1toRef" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes series and compare this with reference DataFrame, as a result gives
param dataset: given series
param ranges: mm : min-max or qq : q1-q3
param  heatmap=False: Dataframe with(reference_above_experiment minimum etc.): rae_min, rae_max, ear_min, ear_max;</p>
<blockquote>
<div>heatmap=True: Series of differences to plot heatmap</div></blockquote>
<p>param relative: only for heatmap, recalculates differences according to the peak size. Warning: negative values are in range -1 to 0
but positive are from 0 to values higher than 1
param reference: path to reference plot
:return: Dataframe (heatmap=False) or Series (heatmap=True)</p>
</dd></dl>

<dl class="function">
<dt id="gwide.profileAnalyser.compareMoretoRef">
<code class="descclassname">gwide.profileAnalyser.</code><code class="descname">compareMoretoRef</code><span class="sig-paren">(</span><em>dataset=Empty DataFrame Columns: [] Index: []</em>, <em>ranges='mm'</em>, <em>reference='/home/tturowski/notebooks/RDN37_reference_collapsed.csv'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/profileAnalyser.html#compareMoretoRef"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.profileAnalyser.compareMoretoRef" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes Dataframe created by filter_df and compare this with reference DataFrame</p>
</dd></dl>

<dl class="function">
<dt id="gwide.profileAnalyser.filter_df">
<code class="descclassname">gwide.profileAnalyser.</code><code class="descname">filter_df</code><span class="sig-paren">(</span><em>input_df=Empty DataFrame Columns: [] Index: [], let_in=[''], let_out=['wont_find_this_string'], smooth=True, window=10</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/profileAnalyser.html#filter_df"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.profileAnalyser.filter_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dataframe for choosen experiments
param input_df: input dataframe
param let_in: list of words that characterize experiment
param let_out: list of word that disqualify experiments (may remain predefined)
param smooth: apply 10nt smootheninig window
:return: dataframe with &#8216;mean&#8217;, &#8216;median&#8217;, &#8216;min&#8217;, &#8216;max&#8217; and quartiles if more than 2 experiments</p>
</dd></dl>

<dl class="function">
<dt id="gwide.profileAnalyser.plot_ChIP">
<code class="descclassname">gwide.profileAnalyser.</code><code class="descname">plot_ChIP</code><span class="sig-paren">(</span><em>df_sense=Empty DataFrame Columns: [] Index: []</em>, <em>df_anti=Empty DataFrame Columns: [] Index: []</em>, <em>title=None</em>, <em>start=None</em>, <em>stop=None</em>, <em>figsize=(15</em>, <em>6)</em>, <em>ylim=(-0.001</em>, <em>0.001)</em>, <em>s_color='red'</em>, <em>as_color='blue'</em>, <em>h_lines=[]</em>, <em>lc='black'</em>, <em>dpi=150</em>, <em>csv_path='/home/tturowski/notebooks/RDN37_reference_collapsed.csv'</em>, <em>color='green'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/profileAnalyser.html#plot_ChIP"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.profileAnalyser.plot_ChIP" title="Permalink to this definition">¶</a></dt>
<dd><p>Function creates plot similar to box plot: median, 2 and 3 quartiles and min-max range</p>
<dl class="docutils">
<dt>csv_path: str()</dt>
<dd>Path to CRAC or other reference file</dd>
</dl>
<p>title: str()</p>
<p>start: int()</p>
<p>stop: int()</p>
<p>figsize: tuple()</p>
<dl class="docutils">
<dt>ylim: tuple()</dt>
<dd>OY axes lim - def (None,0.01)</dd>
<dt>color: str()</dt>
<dd>plot color</dd>
<dt>h_lines: list()</dt>
<dd>Optional. list() of horizontal lines</dd>
<dt>lc: str()</dt>
<dd>color of horizontal lines</dd>
</dl>
<p>None</p>
</dd></dl>

<dl class="function">
<dt id="gwide.profileAnalyser.plot_as_box_plot">
<code class="descclassname">gwide.profileAnalyser.</code><code class="descname">plot_as_box_plot</code><span class="sig-paren">(</span><em>df=Empty DataFrame Columns: [] Index: []</em>, <em>title=None</em>, <em>start=None</em>, <em>stop=None</em>, <em>figsize=(15</em>, <em>6)</em>, <em>ylim=(None</em>, <em>0.01)</em>, <em>dpi=150</em>, <em>color='green'</em>, <em>h_lines=[]</em>, <em>lc='red'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/profileAnalyser.html#plot_as_box_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.profileAnalyser.plot_as_box_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>plots figure similar to box plot: median, 2 and 3 quartiles and min-max range</p>
<dl class="docutils">
<dt>df <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame</span></dt>
<dd>Dataframe containing following columns:<code class="docutils literal"><span class="pre">`['position']</span> <span class="pre">['nucleotide']</span> <span class="pre">['mean']</span> <span class="pre">['median']</span> <span class="pre">['std']`</span></code>
optionally <code class="docutils literal"><span class="pre">`['q1']</span> <span class="pre">['q3']</span> <span class="pre">['max']</span> <span class="pre">['min']`</span></code></dd>
</dl>
<p>title : str</p>
<p>start : int</p>
<p>stop : int</p>
<p>figsize : (float, float)</p>
<dl class="docutils">
<dt>ylim <span class="classifier-delimiter">:</span> <span class="classifier">(float, float)</span></dt>
<dd>OY axes lim. Default = (None,0.01)</dd>
<dt>color <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>line color</dd>
<dt>h_lines <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>optional: list of horizontal lines</dd>
<dt>lc <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>optional: color of horizontal lines</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gwide.profileAnalyser.plot_diff">
<code class="descclassname">gwide.profileAnalyser.</code><code class="descname">plot_diff</code><span class="sig-paren">(</span><em>dataset=Empty DataFrame Columns: [] Index: []</em>, <em>ranges='mm'</em>, <em>label=''</em>, <em>start=None</em>, <em>stop=None</em>, <em>plot_medians=True</em>, <em>plot_ranges=True</em>, <em>figsize=(15</em>, <em>6)</em>, <em>ylim=(None</em>, <em>0.01)</em>, <em>h_lines=[]</em>, <em>reference='/home/tturowski/notebooks/RDN37_reference_collapsed.csv'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/profileAnalyser.html#plot_diff"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.profileAnalyser.plot_diff" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot given dataset and reference, differences are marked
param dataset: dataset from filter_df
param ranges: mm : min-max or qq : q1-q3
param label: label of given dataset
param start: start
param stop: stop
param plot_medians: plot medians
param plot_ranges: plot ranges
param figsize: figzise touple(15, 6)
param ylim: ylim touple(None,0.01)
param h_lines: list of horizontal lines
param reference: path to reference plot
:return: plot with marked differences</p>
</dd></dl>

<dl class="function">
<dt id="gwide.profileAnalyser.plot_from_csv">
<code class="descclassname">gwide.profileAnalyser.</code><code class="descname">plot_from_csv</code><span class="sig-paren">(</span><em>csv_path=''</em>, <em>title=None</em>, <em>start=None</em>, <em>stop=None</em>, <em>figsize=(15</em>, <em>6)</em>, <em>ylim=(None</em>, <em>0.01)</em>, <em>color='green'</em>, <em>h_lines=[]</em>, <em>lc='red'</em>, <em>dpi=75</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/profileAnalyser.html#plot_from_csv"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.profileAnalyser.plot_from_csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots figure similar to box plot: median, 2 and 3 quartiles and min-max range</p>
<dl class="docutils">
<dt>csv_path: str</dt>
<dd>path to csv file</dd>
</dl>
<p>title: str</p>
<p>start: int</p>
<p>stop: int</p>
<dl class="docutils">
<dt>figsize: (float, float)</dt>
<dd>Default = (15,6)</dd>
<dt>ylim <span class="classifier-delimiter">:</span> <span class="classifier">(float, float)</span></dt>
<dd>OY axes lim. Default = (None,0.01)</dd>
<dt>color <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>line color</dd>
<dt>h_lines <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>optional: list of horizontal lines</dd>
<dt>lc <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>optional: color of horizontal lines</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gwide.profileAnalyser.plot_heatmap">
<code class="descclassname">gwide.profileAnalyser.</code><code class="descname">plot_heatmap</code><span class="sig-paren">(</span><em>df=Empty DataFrame Columns: [] Index: []</em>, <em>title='Heat map of differences between dataset and reference plot for RDN37-1'</em>, <em>vmin=None</em>, <em>vmax=None</em>, <em>figsize=(20</em>, <em>10)</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/profileAnalyser.html#plot_heatmap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.profileAnalyser.plot_heatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>plot heat map of differences, from dataframe generated by compare1toRef(dataset, heatmap=True) function</p>
</dd></dl>

<dl class="function">
<dt id="gwide.profileAnalyser.plot_to_compare">
<code class="descclassname">gwide.profileAnalyser.</code><code class="descname">plot_to_compare</code><span class="sig-paren">(</span><em>df=Empty DataFrame Columns: [] Index: []</em>, <em>df2=None</em>, <em>color1='black'</em>, <em>color2='darkred'</em>, <em>label=''</em>, <em>title=None</em>, <em>start=None</em>, <em>stop=None</em>, <em>figsize=(15</em>, <em>6)</em>, <em>ylim=(None</em>, <em>0.01)</em>, <em>h_lines=[]</em>, <em>dpi=75</em>, <em>csv_path='/home/tturowski/notebooks/RDN37_csv_path_collapsed.csv'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/profileAnalyser.html#plot_to_compare"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.profileAnalyser.plot_to_compare" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots given dataset and reference dataset from csv file.</p>
<dl class="docutils">
<dt>df <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame</span></dt>
<dd>Dataframe (dataset) containing following columns:<code class="docutils literal"><span class="pre">`['position']</span> <span class="pre">['nucleotide']</span> <span class="pre">['mean']</span> <span class="pre">['median']</span> <span class="pre">['std']`</span></code>
optionally <code class="docutils literal"><span class="pre">`['q1']</span> <span class="pre">['q3']</span> <span class="pre">['max']</span> <span class="pre">['min']`</span></code></dd>
<dt>df2 <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame</span></dt>
<dd>Optional Dataframe (dataset2). Default = None</dd>
<dt>color1 <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Default color for dataset1 = &#8216;black&#8217;</dd>
<dt>color2 <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Default color for dataset2 = &#8216;darkred&#8217;</dd>
</dl>
<p>label : str</p>
<p>title : str</p>
<p>start : int</p>
<p>stop : int</p>
<p>figsize : (float, float)</p>
<dl class="docutils">
<dt>ylim <span class="classifier-delimiter">:</span> <span class="classifier">(float, float)</span></dt>
<dd>OY axes lim. Default = (None,0.01)</dd>
<dt>h_lines <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>optional: list of horizontal lines</dd>
<dt>dpi <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd>Default dpi=75</dd>
<dt>csv_path <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Default = <code class="docutils literal"><span class="pre">'/home/tturowski/notebooks/RDN37_csv_path_collapsed.csv'</span></code></dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gwide.profileAnalyser.save_csv">
<code class="descclassname">gwide.profileAnalyser.</code><code class="descname">save_csv</code><span class="sig-paren">(</span><em>data_ref=Empty DataFrame Columns: [] Index: []</em>, <em>datasets=Empty DataFrame Columns: [] Index: []</em>, <em>path=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/profileAnalyser.html#save_csv"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.profileAnalyser.save_csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes reference and data DataFrame&#8217;s</p>
<dl class="docutils">
<dt>data_ref <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame</span></dt>
<dd>DataFrame with <code class="docutils literal"><span class="pre">['position']</span></code> and <code class="docutils literal"><span class="pre">['nucleotide']</span></code> columns</dd>
<dt>datasets <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame</span></dt>
<dd>DataFrame containinig experimental data only</dd>
<dt>path <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Optional: path to save csv. Default: None</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">reference DataFrame</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="methods">
<h2>Methods<a class="headerlink" href="#methods" title="Permalink to this headline">¶</a></h2>
<p>Common methods</p>
<span class="target" id="module-gwide.methods"></span><dl class="function">
<dt id="gwide.methods.calGC">
<code class="descclassname">gwide.methods.</code><code class="descname">calGC</code><span class="sig-paren">(</span><em>dataset=Empty DataFrame Columns: [] Index: [], calFor=['G', 'C']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/methods.html#calGC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.methods.calGC" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns GC content in a given dataset
:param dataset: Pandas DataFrame with &#8220;nucleotide&#8221; column
:return: fraction of GC content</p>
</dd></dl>

<dl class="function">
<dt id="gwide.methods.calculateFDR">
<code class="descclassname">gwide.methods.</code><code class="descname">calculateFDR</code><span class="sig-paren">(</span><em>data=Series([]</em>, <em>dtype: float64)</em>, <em>iterations=100</em>, <em>target_FDR=0.05</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/methods.html#calculateFDR"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.methods.calculateFDR" title="Permalink to this definition">¶</a></dt>
<dd><p>calculates False Discovery Rate (FDR) for a given dataset.</p>
<p>data : pd.Series()</p>
<dl class="docutils">
<dt>iterations <span class="classifier-delimiter">:</span> <span class="classifier">int()</span></dt>
<dd>number of iterations. Default = 100</dd>
<dt>target_FDR <span class="classifier-delimiter">:</span> <span class="classifier">float()</span></dt>
<dd>Detault = 0.05</dd>
</dl>
<p>Series()</p>
</dd></dl>

<dl class="function">
<dt id="gwide.methods.cleanNames">
<code class="descclassname">gwide.methods.</code><code class="descname">cleanNames</code><span class="sig-paren">(</span><em>df=Empty DataFrame Columns: [] Index: []</em>, <em>additional_tags=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/methods.html#cleanNames"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.methods.cleanNames" title="Permalink to this definition">¶</a></dt>
<dd><p>Cleans some problems with names if exist</p>
</dd></dl>

<dl class="function">
<dt id="gwide.methods.define_experiments">
<code class="descclassname">gwide.methods.</code><code class="descname">define_experiments</code><span class="sig-paren">(</span><em>paths_in</em>, <em>whole_name=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/methods.html#define_experiments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.methods.define_experiments" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse file names and extract experiment name from them
:param whole_name As defaults script takes first &#8216;a_b_c&#8217; from a_b_c_hittable_reads.txt as experiment name.
:return: list of experiment names and list of paths.</p>
</dd></dl>

<dl class="function">
<dt id="gwide.methods.expNameParser">
<code class="descclassname">gwide.methods.</code><code class="descname">expNameParser</code><span class="sig-paren">(</span><em>name</em>, <em>additional_tags=[]</em>, <em>order='b_d_e_p'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/methods.html#expNameParser"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.methods.expNameParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Function handles experiment name; recognizes AB123456 as experiment date; BY4741 or HTP or given string as bait protein
:param name:
:param additional_tags: list of tags
:param output: default &#8216;root&#8217; ; print other elements when &#8216;all&#8217;
:param order: defoult &#8216;b_d_e_p&#8217; b-bait; d-details, e-experiment, p-prefix
:return: reordered name</p>
</dd></dl>

<dl class="function">
<dt id="gwide.methods.filterExp">
<code class="descclassname">gwide.methods.</code><code class="descname">filterExp</code><span class="sig-paren">(</span><em>datasets, let_in=[''], let_out=['wont_find_this_string']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/methods.html#filterExp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.methods.filterExp" title="Permalink to this definition">¶</a></dt>
<dd><p>for pd.DataFrame() or dict(). Returns object with filtered columns/keys.</p>
<dl class="docutils">
<dt>datasets <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame() or dict()</span></dt>
<dd>DataFrame() or dict() with exp name as a key</dd>
<dt>let_in <span class="classifier-delimiter">:</span> <span class="classifier">list()</span></dt>
<dd>list() with elements of name to filter in</dd>
<dt>let_out <span class="classifier-delimiter">:</span> <span class="classifier">list()</span></dt>
<dd>list() with elements of name to filter out</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">DataFrame() or dict()</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="gwide.methods.getRefFile">
<code class="descclassname">gwide.methods.</code><code class="descname">getRefFile</code><span class="sig-paren">(</span><em>file_from_options</em>, <em>file_type</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/methods.html#getRefFile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.methods.getRefFile" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>Sorting out source of gtf, fasta or tab path, in order (1) from options parser, (2) from ~/bin/default.aml</dt>
<dd>or (3) from environmental variable $xxx_PATH</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>file_from_options</strong> &#8211; file path from options parser, can be an empty string
file_type: &#8216;GTF&#8217;, &#8216;FASTA&#8217;, &#8216;TAB&#8217;</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">path to the file</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="gwide.methods.indexOrder">
<code class="descclassname">gwide.methods.</code><code class="descname">indexOrder</code><span class="sig-paren">(</span><em>df=Empty DataFrame Columns: [] Index: []</em>, <em>additional_tags=[]</em>, <em>output='root'</em>, <em>order='b_d_e_p'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/methods.html#indexOrder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.methods.indexOrder" title="Permalink to this definition">¶</a></dt>
<dd><p>Aplly expNameParser to whole dataframe
:param order: defoult &#8216;b_d_e_p&#8217; b-bait; d-details, e-experiment, p-prefix</p>
</dd></dl>

<dl class="function">
<dt id="gwide.methods.list_paths_in_current_dir">
<code class="descclassname">gwide.methods.</code><code class="descname">list_paths_in_current_dir</code><span class="sig-paren">(</span><em>suffix=''</em>, <em>stdin=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/methods.html#list_paths_in_current_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.methods.list_paths_in_current_dir" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Param:</th><td class="field-body">suffix  -   lists paths in current directory ending with an indicated suffix only
stdin   -   read from standard input instead current directory</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of paths in current dir ending with suffix</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="bigwigtools">
<h2>BigWigTools<a class="headerlink" href="#bigwigtools" title="Permalink to this headline">¶</a></h2>
<p>Tools to analyse BigWig files</p>
<span class="target" id="module-gwide.BigWigTools"></span><dl class="function">
<dt id="gwide.BigWigTools.BigWigCount">
<code class="descclassname">gwide.BigWigTools.</code><code class="descname">BigWigCount</code><span class="sig-paren">(</span><em>datasets={}</em>, <em>genes={}</em>, <em>normalization='pM'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/BigWigTools.html#BigWigCount"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.BigWigTools.BigWigCount" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates coverage for individual genes for given experiments</p>
<dl class="docutils">
<dt>datasets <span class="classifier-delimiter">:</span> <span class="classifier">dict()</span></dt>
<dd>Dictionary with exp name as key and pyBigWig object</dd>
<dt>genes <span class="classifier-delimiter">:</span> <span class="classifier">dict()</span></dt>
<dd>Dictionary with pyBigWig.chroms() dictionary</dd>
<dt>normalization <span class="classifier-delimiter">:</span> <span class="classifier">str()</span></dt>
<dd>Normalization option&#8221; &#8216;&#8220;pM&#8221;, &#8220;pMK&#8221;, None&#8217;. Default: &#8216;&#8220;pM&#8221;&#8217;</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">pd.DataFrame()</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="gwide.BigWigTools.BigWigPileup">
<code class="descclassname">gwide.BigWigTools.</code><code class="descname">BigWigPileup</code><span class="sig-paren">(</span><em>datasets={}</em>, <em>gene_name=''</em>, <em>transcript_length=0</em>, <em>normalization='ntotal'</em>, <em>window=1</em>, <em>median_only=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/BigWigTools.html#BigWigPileup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.BigWigTools.BigWigPileup" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns pd.DataFrame() for multiple experiments with calculated median or all statistic</p>
<dl class="docutils">
<dt>datasets <span class="classifier-delimiter">:</span> <span class="classifier">dict()</span></dt>
<dd>Dictionary with exp name as a key and pyBigWig object</dd>
</dl>
<p>gene_name : str()</p>
<dl class="docutils">
<dt>normalization <span class="classifier-delimiter">:</span> <span class="classifier">str()</span></dt>
<dd>Normalization option&#8221; &#8216;&#8220;pM&#8221;, &#8220;ntotal&#8221;, &#8220;raw&#8221;&#8217;. Default: &#8216;&#8220;ntotal&#8221;&#8217;</dd>
<dt>median_only <span class="classifier-delimiter">:</span> <span class="classifier">boolean</span></dt>
<dd>Default = &#8216;&#8220;False&#8221;&#8217; .If &#8216;&#8220;True&#8221;&#8217; retruns only pd.Series() for median</dd>
<dt>window <span class="classifier-delimiter">:</span> <span class="classifier">int()</span></dt>
<dd>Default = 1. Window size for rolling average to smooth data</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">pd.DataFrame() or pd.Series()</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="gwide.BigWigTools.codons">
<code class="descclassname">gwide.BigWigTools.</code><code class="descname">codons</code><span class="sig-paren">(</span><em>data=Series([]</em>, <em>dtype: float64)</em>, <em>df_genes=Empty DataFrame Columns: [] Index: []</em>, <em>df_sequences=Empty DataFrame Columns: [] Index: []</em>, <em>gene_name=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/BigWigTools.html#codons"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.BigWigTools.codons" title="Permalink to this definition">¶</a></dt>
<dd><p>Annotates data with nucleotide, codone and aminoacid</p>
<p>data : Series()</p>
<dl class="docutils">
<dt>df_genes <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame()</span></dt>
<dd>DataFrame() containig gene_details</dd>
</dl>
<p>gene_name : str()</p>
<dl class="docutils">
<dt>df_sequences <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame()</span></dt>
<dd>sequences</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">DataFrame()</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="gwide.BigWigTools.featureBinCollect">
<code class="descclassname">gwide.BigWigTools.</code><code class="descname">featureBinCollect</code><span class="sig-paren">(</span><em>data=Series([], dtype: float64), df_genes=Empty DataFrame Columns: [] Index: [], gene_name='', bins=[1, 10, 1], warn=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/BigWigTools.html#featureBinCollect"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.BigWigTools.featureBinCollect" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates coverage districution for mRNA elements (UTRs and CDS)</p>
<p>data : Series()</p>
<dl class="docutils">
<dt>df_genes <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame()</span></dt>
<dd>DataFrame() containig gene_details</dd>
</dl>
<p>gene_name : str()</p>
<dl class="docutils">
<dt>bins <span class="classifier-delimiter">:</span> <span class="classifier">list()</span></dt>
<dd>Number of bins for each feature [5UTR, CDS, 3UTR]. Default = &#8216;[1,10,1]&#8217;</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">Series()</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="gwide.BigWigTools.mRNAalign">
<code class="descclassname">gwide.BigWigTools.</code><code class="descname">mRNAalign</code><span class="sig-paren">(</span><em>df_genes=Empty DataFrame Columns: [] Index: []</em>, <em>datasets={}</em>, <em>normalization='ntotal'</em>, <em>window=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/BigWigTools.html#mRNAalign"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.BigWigTools.mRNAalign" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns DataFrame() for multiple experiments with calculated median and etc.</p>
<dl class="docutils">
<dt>df_genes <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame()</span></dt>
<dd>DataFrame() containing genes to align</dd>
<dt>datasets <span class="classifier-delimiter">:</span> <span class="classifier">dict()</span></dt>
<dd>Dictionary with exp name as a key and pyBigWig object</dd>
<dt>normalization <span class="classifier-delimiter">:</span> <span class="classifier">str()</span></dt>
<dd>Normalization option&#8221; &#8216;&#8220;pM&#8221;, &#8220;ntotal&#8221;, &#8220;raw&#8221;&#8217;. Default: &#8216;&#8220;ntotal&#8221;&#8217;</dd>
<dt>window <span class="classifier-delimiter">:</span> <span class="classifier">int()</span></dt>
<dd>Default = 1. Window size for rolling average to smooth data</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">dict() containing DataFrames for different aligners</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="gwide.BigWigTools.metaplot">
<code class="descclassname">gwide.BigWigTools.</code><code class="descname">metaplot</code><span class="sig-paren">(</span><em>df</em>, <em>df2=Empty DataFrame Columns: [] Index: []</em>, <em>h_line=0</em>, <em>title=''</em>, <em>color='black'</em>, <em>lc='red'</em>, <em>figsize=(10</em>, <em>6)</em>, <em>dpi=200</em>, <em>center=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gwide/BigWigTools.html#metaplot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gwide.BigWigTools.metaplot" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates metaplot</p>
<p>df : DataFrame()</p>
<dl class="docutils">
<dt>df2 <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame()</span></dt>
<dd>Dataframe with GC%</dd>
<dt>h_line <span class="classifier-delimiter">:</span> <span class="classifier">int()</span></dt>
<dd>Line visualazing aligning position</dd>
</dl>
<p>title : str()</p>
<p>color : str()</p>
<dl class="docutils">
<dt>lc <span class="classifier-delimiter">:</span> <span class="classifier">str()</span></dt>
<dd>horozontal line color</dd>
</dl>
<p>figsize : touple()</p>
<p>dpi  : int()</p>
<dl class="docutils">
<dt>center <span class="classifier-delimiter">:</span> <span class="classifier">boolean</span></dt>
<dd>detault = True; center plot 1000 bp arount aligning site</dd>
</dl>
</dd></dl>

</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="index.html" class="btn btn-neutral" title="gwide" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Edinburgh Genome Foundry.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>